### ISSUE TITLE
Don't reload the live dev preview when importing files, wait til all files are imported

### ISSUE THREAD

**Issue by flukeout** [mozilla/thimble.mozilla.org#2446]
Was working on a project with a bunch of files, then created a new blank one. In the console it looks like Thimble is still trying to locate / use the files in the previous project...

We discussed this but I'm not sure if we've got it filed so here we go.

<img src="https://user-images.githubusercontent.com/25212/29628486-6fc717a6-87ea-11e7-955c-904b1e4dceec.png" width="350" />

**Comment by gideonthomas** [comment]
I've seen this a few times but not always. If you refresh the page it should go away so I don't think I would classify this as a bug since it's harmless. @humphd might know more about this because I'm pretty sure it's related to how cachestorage works.

**Comment by flukeout** [comment]
This also happens when I drag a bunch of files in for the first time. Check out the screenshot below.

![image](https://user-images.githubusercontent.com/25212/29679315-4b344946-88b6-11e7-854b-dc78d8fa6ab2.png)

This is a project that has a bunch of images files, and as soon as it is imported, there's are two errors in the console for each image that was dragged in.

**Comment by humphd** [comment]
It looks to me like the URL cache didn't get created for some reason.  Why that happened, I don't know, and it would be useful to understand.  If you can recreate this some time, ping me and let's debug together.

I agree that refreshing should fix.  But odd that it happens at all.

**Comment by flukeout** [comment]
@humphd so in looking into this. I can get it to happen every time in Chrome, but not Firefox.
Can you please try the following?

* Download this and project unzip it locally
  * [Archive.zip](https://github.com/mozilla/thimble.mozilla.org/files/1267968/Archive.zip)
* Create a new project in Chrome
* Drag in the files
* That's when I see all the errors.

**Comment by humphd** [comment]
@flukeout yes, I'll test now.

**Comment by flukeout** [comment]
Would also be curious if you see different behaviour between the two browsers.

**Comment by humphd** [comment]
OK, so in Chrome I see this:

![screen shot 2017-08-31 at 1 46 03 pm](https://user-images.githubusercontent.com/427398/29937201-c865742a-8e52-11e7-84d7-21744a571fd4.png)

Which seems very similar to what you described.  I notice that if I pick one of the images that 404s (e.g., `wifi.svg`) and try to open it on its own, I get this:

![screen shot 2017-08-31 at 1 47 11 pm](https://user-images.githubusercontent.com/427398/29937245-ee9a693e-8e52-11e7-9622-a466532c6baf.png)

In other words, it's there in the cache, and the service worker is serving it fine.  As you say, it works in Firefox.

I suspect that there's a timing issue here.  When I refresh the page in Chrome, those 404s are gone.  I think that the file import is triggering a refresh on the new `index.html` page before all the image resources have been loaded/cached.  Maybe the order things get processed is different in Firefox vs. Chrome, so we don't see it there.

We should figure out why the cache isn't getting created properly for all items before we update the preview for an import.

**Comment by humphd** [comment]
Digging into this a bit, here's what's happening.  We drag-and-drop those new project files, which will eventually cause `index.html` to get overwritten.  When that happens (i.e., the file itself gets written to disk, and before the import completes), it will trigger this call stack to occur:

![screen shot 2017-08-31 at 2 04 18 pm](https://user-images.githubusercontent.com/427398/29938056-7a963e84-8e55-11e7-9321-1e72d12b0843.png)
 
When that happens, depending on the order that the files being imported are processed, any dependent resources may or may not have been encountered yet, and may or may not exist in the cache.  If they don't, we'll get those 404s.  However, after the import completes, we trigger a cache refresh and a preview reload, which will mean that the preview will work.

In other words, these 404s are ignorable.  We could fix this by temporarily blocking reloads on the preview, until the import finishes.  That might be a good idea.  We actually already have an API for doing this via the pref used here:

https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/RemoteCommandHandler.js#L95-L102

We could flip that pref off when we start an import, and flip it back on again when we're done.

**Comment by humphd** [comment]
I've changed the title of this bug so it's now about stopping that unnecessary reload during the import.  Whoever wants to work on this, see my comment above.  You'd want to disable auto updates via this:

https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/PostMessageTransport.js#L228-L235

Do that here, before we start the import:

https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/FileImport.js#L58

Then undo it (e.g., return to whatever state the user had originally for auto update) in the callback here:

https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/FileImport.js#L59

**Comment by gideonthomas** [comment]
@humphd does this qualify as a good first bug?

**Comment by humphd** [comment]
Yeah, I think so.

**Comment by ericschv** [comment]
Hi @humphd I would like to tackle this issue.

**Comment by humphd** [comment]
@ericschv OK, perfect, it's yours.  If you follow what I'm discussing in https://github.com/mozilla/thimble.mozilla.org/issues/2446#issuecomment-326378511, I think that will solve this.  NOTE: to fix this bug, and do your dev work, you don't need all of Thimble running, just the Brackets code.  You can run Brackets and then browse to `http://localhost:8000/src/hosted.html` and get the editor that way.

Also, the Thimble devs are in https://chat.mozillafoundation.org/mozilla/channels/thimble, so come say hi.

**Comment by ericschv** [comment]
@humphd thank you. I've just joined the Thimble channel. Looking forward to contributing to this amazing project.

**Comment by ericschv** [comment]
@humphd I'm stuck on trying to call the `setAutoUpdate()` function inside of `brackets/src/filesystem/impls/filer/lib/FileImport.js`. My current attempt at a solution was to import `brackets/src/extensions/default/bramble/lib/RemoteCommandHandler.js` and then call setAutoUpdate() inside of:

https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/FileImport.js#L58

When I import `RemoteCommandHandler.js` the browser is unable to find the other files that are imported within it. From my current understanding the relative paths are conflicting because `FileImport.js` is located within a different directory compared to `RemoteCommandHandler.js`, and as such `FileImport.js` is unable to locate the files that are being imported within `RemoteCommandHandler.js`

Here is a snippet of the error message:

![capture](https://user-images.githubusercontent.com/15128090/32684470-dac106ca-c651-11e7-9256-b6ea5904b55a.JPG)

**Comment by humphd** [comment]
@ericschv I can see why you're having issues here, since the code you're trying to call is crossing the boundary between an extension and the core.  The code in `filesystem/impls/filer/lib/FileImport.js` is part of the core, and anything under `extensions/default/bramble/*` is an extension.  The way that module loading in an extension vs. the core differs, though both use `requirejs`:

* an extension has a `require` context rooted in its extension dir, for example `extensions/default/bramble/`.  Anything you `require()` will get loaded with a path *relative* to that extension dir.
* core modules have a `require`context` rooted in `src/`.  Anything you `require()` will get loaded with a path *relative* to `src/`.

It's possible to mix these two loading systems in some cases.  For example, an extension can load a core module using `brackets.getModule(path)`, which is actually a shortcut to the core `require()` context.

Core modules can't load things in extensions, by design.  You don't want to have the core depend on something optional in an extension (i.e., extensions should be extras that you can enable or disable as you wish).  If you need code to be available to the core, you have to move it into the core.

Probably what we need to do here, is convert [this variable](https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/PostMessageTransport.js#L225-L226) to a preference instead of a module-local variable.  This way, we can change the value of the preference from the core or the extension.

You can see an example of what I mean in how we do JavaScript rewriting.  In the core filesystem code, we have a rewriter module that checks a preference, and only rewrites scripts if it's `true`: https://github.com/mozilla/brackets/blob/master/src/filesystem/impls/filer/lib/HTMLRewriter.js#L143.  We set the value via messages from Thimble in the `RemoteCommandHandler` module, see https://github.com/mozilla/brackets/blob/master/src/extensions/default/bramble/lib/RemoteCommandHandler.js#L104.  The preference is defined here: https://github.com/mozilla/brackets/blob/master/src/editor/Editor.js#L221.  You could probably define it in some part of the `LiveDevelopment/*` code maybe in https://github.com/mozilla/brackets/blob/master/src/LiveDevelopment/LiveDevMultiBrowser.js.

Let me know if this helps, and if you have follow-up questions.  Sorry I missed this earlier in my info above.

**Comment by ericschv** [comment]
@humphd everything in your explanation seems to makes sense up until the part where you recommend defining the preference in `LiveDevelopment/*`. 
  
  What I'm thinking about doing is importing `PreferencesManager` inside of `LiveDevMultiBrowser.js` and then I'm assuming an editor preference variable will have to be declared like so:  `AUTO_UPDATE = "autoUpdate";`. After declaring that variable I plan to define the preference by calling the following function:
```js
PreferencesManager.definePreference(AUTO_UPDATE,         "boolean", true, {
        description: Strings.DESCRIPTION_AUTO_UPDATE
});
``` 
  
  I am not sure where I would call this function inside of `LiveDevMultiBrowser.js`.  Please let me know if I am on the right track.

**Comment by ericschv** [comment]
@humphd I defined the preference in `LiveDevMultiBrowser.js`. Inside of `FileImport.js` I set the preference to false in order to disable reloading live dev preview  before importing, and then set it back to false after importing was finished.
